src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Net.Http.Headers
+{
+    // This struct represents a particular named header --
+    // if the header is one of our known headers, then it contains a reference to the KnownHeader object;
+    // otherwise, for custom headers, it just contains a string for the header name.
+    // Use HeaderDescriptor.Get to resolve an arbitrary header name to a HeaderDescriptor.
+    internal struct HeaderDescriptor : IEquatable<HeaderDescriptor>
+    {
+        private readonly string _headerName;
+        private readonly KnownHeader _knownHeader;
+
+        public HeaderDescriptor(KnownHeader knownHeader)
+        {
+            _knownHeader = knownHeader;
+            _headerName = knownHeader.Name;
+        }
+
+        // This should not be used directly; use static Get below
+        private HeaderDescriptor(string headerName)
+        {
+            _headerName = headerName;
+            _knownHeader = null;
+        }
+
+        public string Name => _headerName;
+        public HttpHeaderParser Parser => _knownHeader?.Parser;
+        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
+
+        public bool Equals(HeaderDescriptor other) =>
+            (_knownHeader == null ?
+                StringComparer.OrdinalIgnoreCase.Equals(_headerName, other._headerName) :
+                _knownHeader == other._knownHeader);
+        public override int GetHashCode() => (_knownHeader == null ? StringComparer.OrdinalIgnoreCase.GetHashCode(_headerName) : _knownHeader.GetHashCode());
+        public override bool Equals(object obj) => (obj is HeaderDescriptor headerInfo ? Equals(headerInfo) : false);
```

Comment:
IIRC, the codegen for is with a struct like this is slightly worse than if you just do it as:
```C#
obj is HeaderDescriptor && Equals((HeaderDescriptor)obj)
```
but my info may be out of date, and regardless it might not matter... hopefully this is never called (in fact, if we expect it's never called, we may just want to throw to ensure we don't accidentally take this path and incur lots of boxing).